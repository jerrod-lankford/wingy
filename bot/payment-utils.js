/* eslint-disable import/prefer-default-export */
import { fryCalc } from './order-utils.js';
import { SIDES, TIP_PERCENT, DELIVERY } from '../common/menu-items.js';

const SMALL_FRY = SIDES.Fries['Regular Fries'];
const LARGE_FRY = SIDES.Fries['Large Fries'];

// Returns both the total and per person values
function calcFries(everyone, person) {
  const fryPeeps = everyone.reduce((accum, curr) => (
    curr.fries === 'Yes' ? accum + 1 : accum
  ), 0);
  const fries = fryCalc(fryPeeps);
  const total = fries.small * SMALL_FRY + fries.large * LARGE_FRY;
  let pp = 0;

  if (person && person.fries === 'Yes') {
    pp = total / fryPeeps;
  }

  return { total, pp };
}

function calcTip(price, fries) {
  return (price + fries) * TIP_PERCENT;
}

function calcDelivery(numPeople) {
  return DELIVERY / numPeople;
}

function calcTax(subtotal, price, fries, tax) {
  const taxPercent = (price + fries) / subtotal;
  return tax * taxPercent;
}

function calcSubTotal(everyone) {
  const fryCost = calcFries(everyone).total;
  const subTotal = everyone.reduce((accum, curr) => (
    accum + curr.price
  ), fryCost);

  return subTotal;
}

export function generatePayment(everyone, totalTax, isDelivery) {
  const subTotal = calcSubTotal(everyone);

  return everyone.map((person) => {
    const { name, price, userId } = person;
    const fries = calcFries(everyone, person).pp;
    const tax = calcTax(subTotal, price, fries, totalTax);
    let tip = 0;
    let delivery = 0;

    if (isDelivery) {
      // Im not being cheap, theres literally no way to tip for pickup
      tip = calcTip(price, fries);
      delivery = calcDelivery(everyone.length);
    }
    const total = price + fries + tax + tip + delivery;

    return {
      name,
      userId,
      price,
      fries,
      tax,
      tip,
      delivery,
      total,
    };
  });
}
