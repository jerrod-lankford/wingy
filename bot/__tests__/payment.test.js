import { generatePayment } from '../payment-utils.js';
import orders from './mockOrders.json';

describe('generate payment for delivery', (everyone) => {
    it.each([
        ['single order no fries'],
        ['single order with fries'],
        ['varying orders'],
        ['order with large fries'],
        ['order with 4 fries'],
        ['order with 5 fries']
    ])('%s', name => {
        const everyone = orders[name];
        const payments = generatePayment(everyone, 2.00, true);
        expect(payments).toMatchSnapshot();
    });
});

describe('generate payment for pickup', (everyone) => {
    it.each([
        ['single order no fries'],
        ['single order with fries'],
        ['varying orders'],
        ['order with large fries'],
        ['order with 4 fries'],
        ['order with 5 fries']
    ])('%s', name => {
        const everyone = orders[name];
        const payments = generatePayment(everyone, 2.00, false);
        expect(payments).toMatchSnapshot();
    });
});