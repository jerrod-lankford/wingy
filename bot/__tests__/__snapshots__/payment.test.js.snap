// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generate payment for delivery order with 4 fries 1`] = `
Array [
  Object {
    "delivery": 0.875,
    "fries": 2.745,
    "name": undefined,
    "price": 9.99,
    "tax": 0.44731296101159107,
    "tip": 1.9102499999999998,
    "total": 15.96756296101159,
    "userId": undefined,
  },
  Object {
    "delivery": 0.875,
    "fries": 2.745,
    "name": undefined,
    "price": 12.99,
    "tax": 0.5526870389884088,
    "tip": 2.3602499999999997,
    "total": 19.52293703898841,
    "userId": undefined,
  },
  Object {
    "delivery": 0.875,
    "fries": 2.745,
    "name": undefined,
    "price": 9.99,
    "tax": 0.44731296101159107,
    "tip": 1.9102499999999998,
    "total": 15.96756296101159,
    "userId": undefined,
  },
  Object {
    "delivery": 0.875,
    "fries": 2.745,
    "name": undefined,
    "price": 12.99,
    "tax": 0.5526870389884088,
    "tip": 2.3602499999999997,
    "total": 19.52293703898841,
    "userId": undefined,
  },
]
`;

exports[`generate payment for delivery order with 5 fries 1`] = `
Array [
  Object {
    "delivery": 0.7,
    "fries": 3.0940000000000003,
    "name": undefined,
    "price": 9.99,
    "tax": 0.35162590701424346,
    "tip": 1.9626,
    "total": 16.098225907014243,
    "userId": undefined,
  },
  Object {
    "delivery": 0.7,
    "fries": 3.0940000000000003,
    "name": undefined,
    "price": 12.99,
    "tax": 0.43224939532383766,
    "tip": 2.4126,
    "total": 19.628849395323837,
    "userId": undefined,
  },
  Object {
    "delivery": 0.7,
    "fries": 3.0940000000000003,
    "name": undefined,
    "price": 9.99,
    "tax": 0.35162590701424346,
    "tip": 1.9626,
    "total": 16.098225907014243,
    "userId": undefined,
  },
  Object {
    "delivery": 0.7,
    "fries": 3.0940000000000003,
    "name": undefined,
    "price": 12.99,
    "tax": 0.43224939532383766,
    "tip": 2.4126,
    "total": 19.628849395323837,
    "userId": undefined,
  },
  Object {
    "delivery": 0.7,
    "fries": 3.0940000000000003,
    "name": undefined,
    "price": 12.99,
    "tax": 0.43224939532383766,
    "tip": 2.4126,
    "total": 19.628849395323837,
    "userId": undefined,
  },
]
`;

exports[`generate payment for delivery order with large fries 1`] = `
Array [
  Object {
    "delivery": 1.1666666666666667,
    "fries": 3.3266666666666667,
    "name": undefined,
    "price": 9.99,
    "tax": 0.6837826272999572,
    "tip": 1.9974999999999998,
    "total": 17.16461596063329,
    "userId": undefined,
  },
  Object {
    "delivery": 1.1666666666666667,
    "fries": 3.3266666666666667,
    "name": undefined,
    "price": 12.99,
    "tax": 0.837826272999572,
    "tip": 2.4475,
    "total": 20.768659606332907,
    "userId": undefined,
  },
  Object {
    "delivery": 1.1666666666666667,
    "fries": 3.3266666666666667,
    "name": undefined,
    "price": 5.99,
    "tax": 0.47839109970047067,
    "tip": 1.3975,
    "total": 12.359224433033804,
    "userId": undefined,
  },
]
`;

exports[`generate payment for delivery single order no fries 1`] = `
Array [
  Object {
    "delivery": 3.5,
    "fries": 0,
    "name": undefined,
    "price": 9.99,
    "tax": 2,
    "tip": 1.4985,
    "total": 16.988500000000002,
    "userId": undefined,
  },
]
`;

exports[`generate payment for delivery single order with fries 1`] = `
Array [
  Object {
    "delivery": 3.5,
    "fries": 4.49,
    "name": undefined,
    "price": 9.99,
    "tax": 2,
    "tip": 2.172,
    "total": 22.152,
    "userId": undefined,
  },
]
`;

exports[`generate payment for delivery varying orders 1`] = `
Array [
  Object {
    "delivery": 0.875,
    "fries": 2.745,
    "name": undefined,
    "price": 12.99,
    "tax": 0.5525899912203687,
    "tip": 2.3602499999999997,
    "total": 19.52283999122037,
    "userId": undefined,
  },
  Object {
    "delivery": 0.875,
    "fries": 0,
    "name": undefined,
    "price": 6.49,
    "tax": 0.22791922739244952,
    "tip": 0.9735,
    "total": 8.56641922739245,
    "userId": undefined,
  },
  Object {
    "delivery": 0.875,
    "fries": 0,
    "name": undefined,
    "price": 14.49,
    "tax": 0.5088674275680422,
    "tip": 2.1734999999999998,
    "total": 18.04736742756804,
    "userId": undefined,
  },
  Object {
    "delivery": 0.875,
    "fries": 2.745,
    "name": undefined,
    "price": 17.49,
    "tax": 0.7106233538191395,
    "tip": 3.03525,
    "total": 24.855873353819142,
    "userId": undefined,
  },
]
`;

exports[`generate payment for pickup order with 4 fries 1`] = `
Array [
  Object {
    "delivery": 0,
    "fries": 2.745,
    "name": undefined,
    "price": 9.99,
    "tax": 0.44731296101159107,
    "tip": 0,
    "total": 13.18231296101159,
    "userId": undefined,
  },
  Object {
    "delivery": 0,
    "fries": 2.745,
    "name": undefined,
    "price": 12.99,
    "tax": 0.5526870389884088,
    "tip": 0,
    "total": 16.28768703898841,
    "userId": undefined,
  },
  Object {
    "delivery": 0,
    "fries": 2.745,
    "name": undefined,
    "price": 9.99,
    "tax": 0.44731296101159107,
    "tip": 0,
    "total": 13.18231296101159,
    "userId": undefined,
  },
  Object {
    "delivery": 0,
    "fries": 2.745,
    "name": undefined,
    "price": 12.99,
    "tax": 0.5526870389884088,
    "tip": 0,
    "total": 16.28768703898841,
    "userId": undefined,
  },
]
`;

exports[`generate payment for pickup order with 5 fries 1`] = `
Array [
  Object {
    "delivery": 0,
    "fries": 3.0940000000000003,
    "name": undefined,
    "price": 9.99,
    "tax": 0.35162590701424346,
    "tip": 0,
    "total": 13.435625907014243,
    "userId": undefined,
  },
  Object {
    "delivery": 0,
    "fries": 3.0940000000000003,
    "name": undefined,
    "price": 12.99,
    "tax": 0.43224939532383766,
    "tip": 0,
    "total": 16.516249395323836,
    "userId": undefined,
  },
  Object {
    "delivery": 0,
    "fries": 3.0940000000000003,
    "name": undefined,
    "price": 9.99,
    "tax": 0.35162590701424346,
    "tip": 0,
    "total": 13.435625907014243,
    "userId": undefined,
  },
  Object {
    "delivery": 0,
    "fries": 3.0940000000000003,
    "name": undefined,
    "price": 12.99,
    "tax": 0.43224939532383766,
    "tip": 0,
    "total": 16.516249395323836,
    "userId": undefined,
  },
  Object {
    "delivery": 0,
    "fries": 3.0940000000000003,
    "name": undefined,
    "price": 12.99,
    "tax": 0.43224939532383766,
    "tip": 0,
    "total": 16.516249395323836,
    "userId": undefined,
  },
]
`;

exports[`generate payment for pickup order with large fries 1`] = `
Array [
  Object {
    "delivery": 0,
    "fries": 3.3266666666666667,
    "name": undefined,
    "price": 9.99,
    "tax": 0.6837826272999572,
    "tip": 0,
    "total": 14.000449293966623,
    "userId": undefined,
  },
  Object {
    "delivery": 0,
    "fries": 3.3266666666666667,
    "name": undefined,
    "price": 12.99,
    "tax": 0.837826272999572,
    "tip": 0,
    "total": 17.154492939666238,
    "userId": undefined,
  },
  Object {
    "delivery": 0,
    "fries": 3.3266666666666667,
    "name": undefined,
    "price": 5.99,
    "tax": 0.47839109970047067,
    "tip": 0,
    "total": 9.795057766367137,
    "userId": undefined,
  },
]
`;

exports[`generate payment for pickup single order no fries 1`] = `
Array [
  Object {
    "delivery": 0,
    "fries": 0,
    "name": undefined,
    "price": 9.99,
    "tax": 2,
    "tip": 0,
    "total": 11.99,
    "userId": undefined,
  },
]
`;

exports[`generate payment for pickup single order with fries 1`] = `
Array [
  Object {
    "delivery": 0,
    "fries": 4.49,
    "name": undefined,
    "price": 9.99,
    "tax": 2,
    "tip": 0,
    "total": 16.48,
    "userId": undefined,
  },
]
`;

exports[`generate payment for pickup varying orders 1`] = `
Array [
  Object {
    "delivery": 0,
    "fries": 2.745,
    "name": undefined,
    "price": 12.99,
    "tax": 0.5525899912203687,
    "tip": 0,
    "total": 16.28758999122037,
    "userId": undefined,
  },
  Object {
    "delivery": 0,
    "fries": 0,
    "name": undefined,
    "price": 6.49,
    "tax": 0.22791922739244952,
    "tip": 0,
    "total": 6.71791922739245,
    "userId": undefined,
  },
  Object {
    "delivery": 0,
    "fries": 0,
    "name": undefined,
    "price": 14.49,
    "tax": 0.5088674275680422,
    "tip": 0,
    "total": 14.998867427568042,
    "userId": undefined,
  },
  Object {
    "delivery": 0,
    "fries": 2.745,
    "name": undefined,
    "price": 17.49,
    "tax": 0.7106233538191395,
    "tip": 0,
    "total": 20.94562335381914,
    "userId": undefined,
  },
]
`;
